Addition of R script to compare CNV caller output

cbethell opened the pull request:
#### Purpose/implementation

The purpose of this draft PR is to produce a copy number plot showing the recurrently amplified/deleted regions detected by CNVkit and Control-FREEC. 

The initial plan was to implement the [TAGCNA](https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0041082&type=printable) workflow, but I encountered some issues when attempting to format the data to fit their pipeline. 

The [`GenVisR` package](https://bioconductor.org/packages/release/bioc/vignettes/GenVisR/inst/doc/Intro.html) was used as a temporary alternative to TAGCNA and the plots produced via this package include:
`analyses/cnv-comparison/plots/compare_cnv_output_frequency.pdf`
`analyses/cnv-comparison/plots/compare_cnv_output_proportion.pdf`

The `ggplot2` package was used to produce the remaining plots found in the `plots` directory of this module. 

#### Issue

This PR addresses issue #8 

#### Directions for reviewers

Per [the comment made] on the relevant issue #8, I attempted to implement the TAGCNA workflow. However, it was unsuccessful due to the specificity of the input format it expects. 

I am still looking into the best way to produce figures like those in the [TAGCNA documentation](https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0041082&type=printable), but felt it necessary to show what has been produced thus far. 

1. Given the purpose of this analysis, I chose to write it in a R script. I used the `sample-distribution-analysis` module as a guide in this sense. That being said, would anyone argue for the need of separate scripts and a shell script to sequentially execute said scripts?

2. I also began to include `optparse` functions and decided to file a draft PR of the script as is so that I can receive a second opinion on whether or not command line options are necessary for this module (again, I used the `sample-distribution-analysis` module as a reference).

#### Results

This script produces five plots as follows:
`analyses/cnv-comparison/plots/compare_cnv_output_barplot_aberration.pdf`
`analyses/cnv-comparison/plots/compare_cnv_output_barplot_histology.pdf`
`analyses/cnv-comparison/plots/compare_cnv_output_boxplot.pdf`
`analyses/cnv-comparison/plots/compare_cnv_output_frequency.pdf`
`analyses/cnv-comparison/plots/compare_cnv_output_proportion.pdf`

#### Docker and continuous integration

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### PR Checklist
* [x] Run a linter (not done as yet)
* [x] Set the seed (if applicable)
* [x] Comments and/or documentation up to date
* [x] Double check your paths

cbethell commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,200 @@
+  # Given the file path of the CNV data, read in the data.
+  #
+  # Arg:
+  #   input_directory: file path of input directory
```
Good point, I do see how this can be confusing. I am sure this was better suited for my original, more complicated script, but making it one `file.path` to the data does seem to make the most sense now.

cbethell commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,200 @@
+  # Given the data.frame containing CNV caller output, return a data.frame with
+  # a column with the values 0 and 1 to represent loss and gain, respectively.
+  #
+  # Args:
```
I like this idea.

cbethell commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,200 @@
+    read.table(gzfile(file.path(input_directory, file_path)), header = TRUE)
+
+  # Rename the columns
+  dataframe <-
```
I now realize that my comment above this line was misleading as this was meant to rearrange the columns, rather than simply rename them. That being said, the function `colnames` won't be feasible here. I am thinking of using `dplyr::select` as the replacement here unless either of you may have a better solution.

cansav09 replied:
Yeah, do `dplyr::select/rename`

cbethell commented:
@jashapiro I'll wait on your review before changing this script into a R notebook or running it from a bash script with `optparse` options. I have committed changes that address all of the other comments @cansav09 made.

cbethell commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,200 @@
+
+  # Rearrange the columns
+  cnv_df <- cnv_df %>%
+    dplyr::select("chrom", "loc.start", "loc.end", "ID", "num.mark", "seg.mean")
```
Good point, will do.

cbethell commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,200 @@
+
+}
+
+plot_cn_freq <-
```
I also see your point here, however, the motivation behind making this a function was to keep the calling of functions in the `01-cnv-comparison-plotting.R` script uniformed and the number of lines of code to a minimum. 

That being said, would your recommendation here be to call the `GenVisR` function in the `01` script nonetheless?

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,171 @@
+    dplyr::select("chrom", "loc.start", "loc.end", "ID", "num.mark", "seg.mean")
+  
+  # Reorder chromosome levels 
+  levels(cnv_df$chrom) <- c(paste0("chr", c(1:22, "X", "Y")))
```
The above doesn't actually convert to a factor, but makes it a kind of strange hybrid character with a levels attribute:
`dplyr::mutate(chrom = factor(chrom, levels = paste0("chr", c(1:22, "X", "Y"))`
should work

jashapiro replied:
If you look at the before and after plots, you can see that the order didn't actually change but the labels did. Which would be bad...

cbethell replied:
Ahh, I thought it was strange when I initially tried to convert to a factor, what R claimed to already be a factor. Thanks, I'll give that a shot now.

jashapiro replied:
oooh... add stringsAsFactors=False to your read.table (or use readr...)

cbethell replied:
I also made that realization a minute ago. That'll also be in this upcoming commit 👍

jashapiro commented on analyses/cnv-comparison/01-cnv-comparison-plotting.R:
```diff
@@ -0,0 +1,123 @@
+
+if (!("cowplot" %in% installed.packages())) {
+  install.packages("BiocManager")
+  BiocManager::install("cowplot")
```
Is there a reason not tin install this from cran?

cbethell replied:
Good question, I cannot pinpoint my initial motivation here so I can change this to install from cran.

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,169 @@
+  meta_joined <- filtered_cnv_df %>%
+    dplyr::inner_join(metadata, by = c("sample" = "Kids_First_Biospecimen_ID"))
+
+  # Create barplot where the y-axis represents the size of aberration
```
On second look, I am pretty confused by these plots. What is actually being plotted here? I had thought it was the number of each type in each chromosome, but the `stat = "identity"` is confusing me. Would the default `stat = "count"` make more sense? 

On some investigation, it looks like these ultimately produce the same plots as the ones that I was expecting, but if you leave off the `y` aesthetic, and use the default `geom_bar()` (no stat) you get a much smaller file size and better y axis labels.

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,169 @@
+  # Create barplot where the y-axis represents the size of aberration
+  barplot <- ggplot2::ggplot(filtered_cnv_df,
+                             ggplot2::aes(x = chromosome,
+                                          y = as.factor(aberration),
```
See earlier comment about the y axis/stat

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,169 @@
+                                          fill = broad_histology)) +
+    ggplot2::geom_bar(stat = "identity") +
+    ggplot2::theme_bw() +
+    ggplot2::theme(axis.text.y = ggplot2::element_blank())
```
Add y axis label?

jharenza commented:
> This looks pretty good now, though I am very much confused by the results: the two callers are kind of shockingly different, and I find it very strange that ControlFREEC is calling more on the smaller chromosomes. That result makes me somewhat suspicious about what is going on, and I would like to hear more about the data there.
> 
> There is one major change that I think would be good for the ggplot-based plots, and that is to combine the two data frames with an extra column designating whether the sample is from CNVkit vs ControlFREEC, then using faceting to split the two, rather than relying on cowplot. This would have the advantage of putting the two plots on the same scale. You can do this pretty easily with `dplyr::bind_rows()` using the `.id` option to specify the new column name.
> 
> I would also like to see the results for the histology plot separated by aberration type, especially since the two methods have such different proportions of gain/loss calls.
> 
> On what I think was one of the original questions, I do think this script might probably be better off converted to a notebook, to give some space to comment a bit on these results, rather than just producing the raw plots.

@cbethell and @jashapiro - would love to see this in a notebook. Of note, we are still benchmarking  (nearing the end) the CN algorithms (and have a caveat on this in the README). We had a few samples show genome-wide gain and after tweaking parameters and adding options for BAF adjustment, we have much better results. Of additional note, I thought I mentioned this, but forgive me if not, I think we would ideally use results from #128 for this ticket, though I think @cgreene and I thought this ticket could still be worked on to get the code ready for a later plug and chug. We have not created a PR for #128 yet because we realized we had to use different CN files from what we had released (containing predicted gain/loss info without the need to threshold LRR) in order to create consensus (these data will be added, hopefully in #146), but the early data is only showing ~50% consensus. We are now working on integrating Manta as a 3rd algorithm and doing a 2/3 consensus and want to toy with degree of overlap based on size of CNV, rather than 2/2. Long story short, I can help you review the results by histology if in a notebook based on what we know about certain diseases and CNV expectations!

jaclyn-taroni commented:
>  Of additional note, I thought I mentioned this, but forgive me if not, I think we would ideally use results from #128 for this ticket, though I think @cgreene and I thought this ticket could still be worked on to get the code ready for a later plug and chug. 

From https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/8#issuecomment-535018323: 
> I imagine the output would still be a SEG file, such that you can start with any of the SEGs we provide.

This appears to have the functionality to read in files in that format:  https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/142/files#diff-fd9d53007cdfd11cf0ae23e3eede7eabR12

> Long story short, I can help you review the results by histology if in a notebook based on what we know about certain diseases and CNV expectations!

What are you looking for in a notebook @jharenza? Do you want to see the plot here: [`cnv-comparison/plots/compare_cnv_output_barplot_histology.pdf`](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/6222c9589bcc68153b54d2598930707644e0d47f/analyses/cnv-comparison/plots/compare_cnv_output_barplot_histology.pdf) or are you looking for additional analyses? My inclination is to, once checked for correctness, get this merged and iterate on it.

jharenza commented:
@jaclyn-taroni ahh thanks for the link, i didn't readily see the plot - the notebook idea was just to see everything gathered together for quick visuals. My question for this analysis - is this more of a QC or do we want to turn this into a manuscript figure? If the latter, I think having this plotted in GISTIC-type format for the different histologies or as an IGV-type figure, grouped by histology, would be most effective for showing recapitulation of previous findings.

cbethell commented:
> @jaclyn-taroni ahh thanks for the link, i didn't readily see the plot - the notebook idea was just to see everything gathered together for quick visuals. My question for this analysis - is this more of a QC or do we want to turn this into a manuscript figure? If the latter, I think having this plotted in GISTIC-type format for the different histologies or as an IGV-type figure, grouped by histology, would be most effective for showing recapitulation of previous findings.

@jharenza I can work on plotting this as an IGV-type figure. My initial attempt to apply GISTIC here was unsuccessful due to compatibility issues with my OS. However, I can give it a more thorough attempt or explore other avenues to produce this figure. I will mention you on this issue for further review once produced.

jharenza commented:
OK - I can run GISTIC quickly on the v5 data and potentially add this to the data release, will send you the files so you can get started. I also noticed maftools updated some of their CN plotting abilities, so producing those per histology could be nice, too.

cbethell commented:
@jashapiro I addressed the comments you made above. Does this look ready to merge now? The changes involving GISTIC can be made and filed in a separate PR.

jashapiro commented on analyses/cnv-comparison/01-cnv-comparison-plotting.Rmd:
```diff
@@ -0,0 +1,178 @@
+cnv_list <-
+  list(
+    file.path(input_directory, "pbta-cnv-cnvkit.seg.gz"),
+    file.path(input_directory, "pbta-cnv-controlfreec.seg.gz")
```
Name here to save work later.
```suggestion
    cnvkit = file.path(input_directory, "pbta-cnv-cnvkit.seg.gz"),
    controlfreec = file.path(input_directory, "pbta-cnv-controlfreec.seg.gz")
```

jaclyn-taroni replied:
Use `<-` assignment though please 🙂

jashapiro replied:
not assignment, naming!

jaclyn-taroni replied:
Nope I'm wrong!

jaclyn-taroni replied:
🎢

jashapiro commented on analyses/cnv-comparison/01-cnv-comparison-plotting.Rmd:
```diff
@@ -0,0 +1,178 @@
+# Bind rows of dataframes in cnv_filtered
+combined_cnv_filtered <- dplyr::bind_rows(cnv_filtered, .id = "cnv_caller") 
+combined_cnv_filtered$cnv_caller <- gsub("1", "cnvkit", combined_cnv_filtered$cnv_caller)
+combined_cnv_filtered$cnv_caller <- gsub("2", "controlfreec", combined_cnv_filtered$cnv_caller)
```
With the named list above, this should be correct automatically, so you should be able to delete these two lines.

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,179 @@
+                                          fill = broad_histology)) +
+    ggplot2::geom_bar() +
+    ggplot2::theme_bw() +
+    ggplot2::facet_grid(cnv_caller~aberration) +
```
```suggestion
    ggplot2::facet_grid(cnv_caller ~ aberration) +
```

jashapiro commented on analyses/cnv-comparison/util/cnv-comparison-functions.R:
```diff
@@ -0,0 +1,179 @@
+                                          fill = aberration)) +
+    ggplot2::geom_bar() +
+    ggplot2::theme_bw() +
+    ggplot2::facet_grid(cnv_caller~aberration) +
```
```suggestion
    ggplot2::facet_grid(cnv_caller ~ aberration) +
```

cbethell commented:
> Looks good! (Except for ControlFreec...)

Thank you and agreed!
